# Task 1 Full Results Summary

This notebook summarizes the full-mode metrics of the baseline model and the three ablation studies for basis function 04.

Experiments included:
- full_model
- ablation_no_fno
- ablation_no_attention
- ablation_no_gradient_loss

In [1]:
from pathlib import Path
import pandas as pd

metric_files = {
    "full_model": Path("outputs/teaching_single_basis/basis_04/metrics_basis_04.csv"),
    "ablation_no_fno": Path("outputs/ablation_no_fno/basis_04/metrics_basis_04.csv"),
    "ablation_no_attention": Path("outputs/ablation_no_attention/basis_04/metrics_basis_04.csv"),
    "ablation_no_gradient_loss": Path("outputs/ablation_no_gradient_loss/basis_04/metrics_basis_04.csv"),
}

rows = []

for name, path in metric_files.items():
    print(name, path.exists(), path)
    if path.exists():
        df = pd.read_csv(path)
        df.insert(0, "experiment", name)
        rows.append(df)

comparison = pd.concat(rows, ignore_index=True)
comparison.to_csv("outputs/task1_full_metrics_comparison.csv", index=False)
comparison

full_model True outputs/teaching_single_basis/basis_04/metrics_basis_04.csv
ablation_no_fno True outputs/ablation_no_fno/basis_04/metrics_basis_04.csv
ablation_no_attention True outputs/ablation_no_attention/basis_04/metrics_basis_04.csv
ablation_no_gradient_loss True outputs/ablation_no_gradient_loss/basis_04/metrics_basis_04.csv


,experiment,basis_id,mse,rmse,mae,r2,relative_l2,grad_mse,max_abs_error,n_samples
0,full_model,4,0.012038,0.109716,0.037017,0.931886,0.260202,0.041623,1.784681,100
1,ablation_no_fno,4,0.009339,0.096638,0.039342,0.947134,0.228946,0.031159,1.793450,100
2,ablation_no_attention,4,0.013008,0.114052,0.039226,0.926424,0.272993,0.046128,1.791959,100
3,ablation_no_gradient_loss,4,0.012894,0.113552,0.038463,0.927099,0.269931,0.045090,1.785915,100


## Full-mode comparison summary

The table above summarizes the full-mode results for basis function 04 across the baseline model and three ablation studies.

### Main observations

* The **no-FNO** ablation achieved the best overall performance among the four runs.
* It obtained the lowest **MSE** (`0.009339`), lowest **RMSE** (`0.096638`), highest **R2** (`0.947134`), lowest **Relative L2** (`0.228946`), and lowest **Gradient MSE** (`0.031159`).
* The **full model** performed strongly, but it did not outperform the no-FNO variant on this basis-04 experiment.
* The **no-attention** and **no-gradient-loss** variants both performed worse than the full model on most metrics.
* This suggests that **attention gates** and **gradient matching loss** both contribute positively to performance for basis function 04.
* However, the result of the **no-FNO** ablation shows that the FNO-style preprocessing block is not guaranteed to improve performance in every case.

### Interpretation

For this specific task setting, removing the FNO-style preprocessing block unexpectedly improved the final metrics, while removing attention gates or removing gradient loss degraded performance. This indicates that the contribution of each module can depend on the target basis function, the optimization dynamics, and the interaction between modules.

### Takeaway

This ablation study shows why controlled comparison is important: a module that seems theoretically useful does not always improve empirical performance, while other components such as attention and gradient-aware supervision may provide more consistent benefit in this setting.
